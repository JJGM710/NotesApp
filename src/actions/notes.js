import { db } from '../firebase/firebaseConfig';
import { types } from './../types/types';
import { loadNotes } from './../helpers/loadNotes';
import Swal from 'sweetalert2';
import { loadEntries } from './../helpers/loadEntries';
import { fileUpload } from './../helpers/fileUpload';

//accion asincrona que crea una nueva nota y luego hace el dispatch para que se active una nota, y se empiezan
// a cargar las nuevas entradas
export const startNewNote = () => {
	return async (dispatch, getState) => {
		const uid = getState().auth.uid;
		const newNote = {
			title: '',
			body: '',
			date: new Date().getTime(),
			url: null,
			update: null,
		};

		try {
			const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
			dispatch(activeNote(doc.id, newNote));
			dispatch(startLoadingEntries(uid));
			dispatch(addNote(doc.id, newNote));
		} catch (error) {
			console.log(error);
		}

		// console.log(doc);
	};
};

//Carga las notas  y las coloca en el state del redux
export const startLoadingNotes = (uid) => {
	return async (dispatch, getState) => {
		const note = await loadNotes(uid);
		dispatch(setNotes(note));
	};
};

//carga el numero de entraddas que hay direectmaen desde la base de datos
//se puede optimizar para que sea lalmada de StartLoadingNotes
export const startLoadingEntries = (uid) => {
	return async (dispatch, getState) => {
		const entradas = await loadEntries(uid);
		dispatch(setEntry(entradas));
		console.log(entradas);
	};
};

//guarda la nota y hace el dispatch de refresh note para cambiar las propiedades de esa nota modificada
export const startSaveNote = (note) => {
	return async (dispatch, getState) => {
		const uid = getState().auth.uid;
		const { active: activeNote } = getState().notes;

		if (!note.url) {
			delete note.url;
		}

		const noteToFirestore = { ...note };
		delete noteToFirestore.id;
		//asignacion de propiedad para update y mandar a actualizar en la base de datos
		noteToFirestore.update = new Date().getTime();

		await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
		//asignacion a la variable local luego de que se realizo el save en la base de datos
		activeNote.update = noteToFirestore.update;

		//se hace el refresh para que las  notas se actualicen con los nuevos cambios

		dispatch(refreshNote(note.id, noteToFirestore));
		Swal.fire('Saved', note.title, 'success');
	};
};

export const startUploading = (file) => {
	return async (dispatch, getState) => {
		const { active: activeNote } = getState().notes;

		Swal.fire({
			title: 'Uploading....',
			text: 'Please wait...',
			allowOutsideClick: false,
			onBeforeOpen: () => {
				Swal.showLoading();
			},
		});

		const fileUrl = await fileUpload(file);

		Swal.close();
		activeNote.url = fileUrl;
		dispatch(startSaveNote(activeNote));
	};
};

export const startDelete = (id) => {
	return async (dispatch, getState) => {
		const uid = getState().auth.uid;

		await db.doc(`${uid}/journal/notes/${id}`).delete();

		dispatch(deleNote(id));
		dispatch(startLoadingEntries(uid));
	};
};

//accciones sincronas
export const activeNote = (id, note) => {
	return {
		type: types.notesActive,
		payload: {
			id,
			...note,
		},
	};
};
export const setNotes = (notes) => {
	return {
		type: types.notesLoad,
		payload: notes,
	};
};

export const refreshNote = (id, note) => {
	return {
		type: types.notesUpdated,
		payload: {
			id,
			note: {
				id,
				...note,
			},
		},
	};
};

export const addNote = (id, note) => {
	return {
		type: types.notesAddNew,
		payload: {
			id,
			...note,
		},
	};
};

export const setEntry = (entradas) => {
	return {
		type: types.notesEntries,
		payload: entradas,
	};
};

export const deleNote = (id) => {
	return {
		type: types.notesDelete,
		payload: id,
	};
};

export const noteLogout = () => {
	return {
		type: types.notesLogoutCleaning,
	};
};
