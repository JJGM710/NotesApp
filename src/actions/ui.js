import { types } from './../types/types';
export const setError = (err) => {
	return {
		type: types.uiSetError,
		payload: err,
	};
};

export const removeError = () => {
	return {
		type: types.uiRemoveError,
	};
};

export const redirectError = () => {
	return {
		type: types.uiRedirect,
	};
};

export const removeRedirectError = () => {
	return {
		type: types.uiRemoveRedirect,
	};
};

export const startLoading = () => {
	return {
		type: types.uiStartLoading,
	};
};

export const finishLoading = () => {
	return {
		type: types.uiFinishLoading,
	};
};
