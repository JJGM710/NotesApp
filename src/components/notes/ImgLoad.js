import React, { useState } from 'react';
import Spinner from 'react-bootstrap/Spinner';

export const ImgLoad = ({ url }) => {
	const [loading, setLoading] = useState(false);
	const handleOnload = () => {
		setLoading(true);
	};

	return (
		<>
			{loading === false && (
				<div>
					<Spinner animation="grow" />
					<Spinner animation="grow" />
					<Spinner animation="grow" />
				</div>
			)}
			<div className="animate__animated animate__fadeIn ">
				<img
					src={url}
					alt="Imagen"
					className="notes__image "
					onLoad={handleOnload}
					style={{ display: loading ? 'block' : 'none' }}
				/>
			</div>
		</>
	);
};
