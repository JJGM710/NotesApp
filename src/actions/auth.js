import { types } from './../types/types';
import { firebase } from '../firebase/firebaseConfig';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import {
	setError,
	finishLoading,
	startLoading,
	removeRedirectError,
} from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

export const startLogin = (email, password) => {
	return (dispatch) => {
		dispatch(startLoading());
		return firebase
			.auth()
			.signInWithEmailAndPassword(email, password)
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
				// dispatch(removeError());
				dispatch(finishLoading());
				Swal.fire('Logeed IN!', 'User Auth Correct', 'success');
			})
			.catch((e) => {
				// dispatch(setError(e.message));
				Swal.fire('Error', e.message, 'error');
				dispatch(finishLoading());
			});
	};
};

export const startRegister = (name, email, password) => {
	return (dispatch) => {
		dispatch(startLoading());
		return firebase
			.auth()
			.createUserWithEmailAndPassword(email, password)
			.then(async ({ user }) => {
				await user.updateProfile({ displayName: name });
				console.log(user.displayName);
				dispatch(login(user.uid, user.displayName));
				dispatch(finishLoading());
			})
			.catch((e) => {
				dispatch(finishLoading());
				Swal.fire('Error', e.message, 'error');
			});
	};
};

export const startGoogleLogin = () => {
	return (dispatch) => {
		firebase
			.auth()
			.signInWithPopup(googleAuthProvider)
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
			});
	};
};

export const login = (uid, displayName) => {
	return {
		type: types.login,
		payload: {
			uid,
			displayName,
		},
	};
};

export const startLogout = () => {
	return async (dispatch) => {
		dispatch(setError('Has cerrado sesion'));
		await firebase.auth().signOut();
		dispatch(removeRedirectError());
		dispatch(logout());
		dispatch(noteLogout());
	};
};

export const logout = () => ({
	type: types.logout,
});
