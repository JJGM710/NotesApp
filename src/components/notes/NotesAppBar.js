import React from 'react';
import Moment from 'react-moment';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from './../../actions/notes';

export const NotesAppBar = () => {
	const dispatch = useDispatch();
	const { active } = useSelector((state) => state.notes);
	// const noteDate = moment(active.update);

	const handleSave = () => {
		// console.log(active);
		dispatch(startSaveNote(active));
	};

	const handlePictureUpload = () => {
		console.log('picture');
		document.querySelector('#fileSelector').click();
		document.querySelector('#fileSelector').value = '';
	};

	const handleFileChange = (e) => {
		const file = e.target.files[0];
		console.log(e);
		if (file) {
			dispatch(startUploading(file));
		}
	};

	return (
		<div className="notes__appbar">
			{active.update && (
				<Moment interval={60000} fromNow>
					{active.update}
				</Moment>
			)}
			{/* <span>{noteDate.fromNow()}</span> */}

			<input
				id="fileSelector"
				type="file"
				name="file"
				style={{ display: 'none' }}
				onChange={handleFileChange}
			/>

			<div>
				<button className="btn btn-info" onClick={handlePictureUpload}>
					Picture
				</button>
				<button className="btn btn-success ml-1" onClick={handleSave}>
					Save
				</button>
			</div>
		</div>
	);
};
