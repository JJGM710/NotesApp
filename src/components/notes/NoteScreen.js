import React, { useEffect, useRef } from 'react';
import { NotesAppBar } from './NotesAppBar';
import { useSelector, useDispatch } from 'react-redux';
import { useForm } from './../../hooks/useForm';
import { activeNote } from '../../actions/notes';
import { startDelete } from './../../actions/notes';
import { ImgLoad } from './ImgLoad';

export const NoteScreen = () => {
	const dispatch = useDispatch();
	const { active } = useSelector((state) => state.notes);
	const { formValues, handleInputChange, reset } = useForm({
		title: active.title,
		body: active.body,
		date: active.date,
		id: active.id,
	});
	const activeID = useRef(active.id);
	const { body, title } = formValues;
	console.log(formValues);

	useEffect(() => {
		if (active.id !== activeID.current) {
			reset({
				title: active.title,
				body: active.body,
				date: active.date,
				id: active.id,
			});
			activeID.current = active.id;
		}
	}, [active, reset]);

	useEffect(() => {
		dispatch(
			activeNote(formValues.id, {
				...formValues,
			})
		);
	}, [formValues, dispatch]);

	const handleDelete = () => {
		dispatch(startDelete(active.id));
	};

	return (
		<div className="notes__main-content animate__animated animate__zoomIn animate__faster">
			<NotesAppBar />

			<div className="notes__content">
				<input
					name="title"
					type="text"
					placeholder="Some awesome title"
					className="notes__title-input"
					value={title}
					onChange={handleInputChange}
				/>
				<textarea
					name="body"
					className="notes__textarea"
					placeholder="What happened today?"
					value={body}
					onChange={handleInputChange}
				></textarea>

				{active.url && <ImgLoad url={active.url} />}
			</div>
			<button className="btn btn-danger" onClick={handleDelete}>
				Delete
			</button>
		</div>
	);
};
