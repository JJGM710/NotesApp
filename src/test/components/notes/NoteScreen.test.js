import React from 'react';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import { activeNote } from '../../../actions/notes';
import { NoteScreen } from './../../../components/notes/NoteScreen';

jest.mock('../../../actions/notes', () => ({
	activeNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
	auth: { uid: '1', name: 'JuanJose' },
	ui: { loading: false, msgError: null },
	notes: {
		notes: [],
		active: { id: 1234, title: 'hola', body: 'body', date: 0 },
		config: 2,
	},
};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
	<Provider store={store}>
		<NoteScreen />
	</Provider>
);
describe('Pruebas en NoteScreen', () => {
	test('debe mostrarse correctamente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe de disparar el active note', () => {
		wrapper.find('input[name="title"]').simulate('change', {
			target: {
				name: 'title',
				value: 'Hola de nuevo',
			},
		});

		expect(activeNote).toHaveBeenLastCalledWith(1234, {
			body: 'body',
			title: 'Hola de nuevo',
			id: 1234,
			date: 0,
		});
	});
});
