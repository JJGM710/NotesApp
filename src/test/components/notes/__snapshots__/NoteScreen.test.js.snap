// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en NoteScreen debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__zoomIn animate__faster"
>
  <div
    className="notes__appbar"
  >
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn btn-info"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn btn-success ml-1"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value="body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
